keeper/msg_server_vaults.go

```diff
@@ -433,3 +437,73 @@ func (k *Keeper) ToUserVaultPositionModuleAccount(address string, vaultType vaul
 		return vaults.StakedVaultName
 	}
 }
+
+// VaultsEndProgram handles the logic to end the vaults program, unlocking all user positions.
+func (k *Keeper) VaultsEndProgram(ctx context.Context) error {
+	// Get all the vaults positions.
+	positions, err := k.GetVaultsPositions(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Create a mapping between the different vault types by address and the total positions amount.
+	stakedUsers := map[string]math.Int{}
+	flexibleUsers := map[string]math.Int{}
+
+	// Iterate through all the positions.
+	k.logger.Info("collecting Positions...")
+	for _, position := range positions {
+		addr, err := k.address.BytesToString(position.Address)
+		if err != nil {
+			k.logger.Warn("invalid address encountered: " + err.Error())
+			continue
+		}
+		if position.Vault == vaults.FLEXIBLE {
+			if _, exists := flexibleUsers[addr]; !exists {
+				flexibleUsers[addr] = position.Amount
+			} else {
+				flexibleUsers[addr] = flexibleUsers[addr].Add(position.Amount)
+			}
+		} else {
+			if _, exists := stakedUsers[addr]; !exists {
+				stakedUsers[addr] = position.Amount
+			} else {
+				stakedUsers[addr] = stakedUsers[addr].Add(position.Amount)
+			}
+		}
+	}
+
+	// First, unlock all the Staked Vault positions.
+	k.logger.Info(fmt.Sprintf("unlocking %d staked vault positions...", len(stakedUsers)))
+	var stakedUsersProcessed int
+	for stakedUserAddr, stakedUserTotalAmount := range stakedUsers {
+		if _, err := k.unlock(ctx, &vaults.MsgUnlock{
+			Signer: stakedUserAddr,
+			Vault:  vaults.STAKED,
+			Amount: stakedUserTotalAmount,
+		}); err != nil {
+			k.logger.Error(fmt.Sprintf("failed to unlock staked vault position for %s: %v", stakedUserAddr, err))
+			continue
+		}
+		stakedUsersProcessed += 1
+	}
+	k.logger.Info(fmt.Sprintf("unlocked %d/%d staked vault positions!", len(stakedUsers), stakedUsersProcessed))
+
+	// Then, unlock all the Flexible Vault positions.
+	k.logger.Info(fmt.Sprintf("unlocking %d flexible vault positions...", len(flexibleUsers)))
+	var flexibleUsersProcessed int
+	for flexibleUserAddr, flexibleUserTotalAmount := range flexibleUsers {
+		if _, err := k.unlock(ctx, &vaults.MsgUnlock{
+			Signer: flexibleUserAddr,
+			Vault:  vaults.FLEXIBLE,
+			Amount: flexibleUserTotalAmount,
+		}); err != nil {
+			k.logger.Error(fmt.Sprintf("failed to unlock flexible vault position for %s: %v", flexibleUserAddr, err))
+			continue
+		}
+		flexibleUsersProcessed += 1
+	}
+	k.logger.Info(fmt.Sprintf("unlocked %d/%d flexible vault positions!", len(flexibleUsers), flexibleUsersProcessed))
```

Comment:
I think this is the wrong way around? 🤔 

```suggestion
	k.logger.Info(fmt.Sprintf("unlocked %d/%d flexible vault positions!", flexibleUsersProcessed, len(flexibleUsers)))
```

Reply:
good catch!